### DISCUSSION TITLE
`model_dump_json` "swallows" content of nested model

### DISCUSSION THREAD

**Discussion by fproske** [pydantic/pydantic#6819] [Question] [answered]
Hi all,

I'm in the process of migrating some of my pydantic models to version 2. While doing so, I stumbled upon the following issue:

```python
from pydantic import BaseModel, InstanceOf

class UserEventModel(BaseModel):
    pass

class TestEventModel(UserEventModel):
    test: str


class WebhookPayload(BaseModel):
    payload: InstanceOf[UserEventModel]


print(WebhookPayload(payload=TestEventModel(test='bla')).model_dump())
print(WebhookPayload(payload=TestEventModel(test='bla')).model_dump_json())
```
While `model_dump` gives me the desired output:
```
{'payload': {'test': 'bla'}}
```
`model_dump_json` swallows the content of my payload:
```
{"payload":{}}
```
Does anyone have an idea what I'm doing wrong here?

Thank you and best regards,
Freddy

**Comment by Kludex** [comment]
How were you doing this on V1?

**Reply by fproske** [reply]
I used to do:

```python
from pydantic import BaseModel

class UserEventModel(BaseModel):
    pass

class TestEventModel(UserEventModel):
    test: str

class WebhookPayload(BaseModel):
    payload: UserEventModel

print(WebhookPayload(payload=TestEventModel(test='bla')).json())
```

Not sure if this was ever an intended use, though. For now, I use `typing.Any` instead of `InstanceOf[UserEventModel]`, which seems to do the trick, but is not as clean imo.

**Comment by dmontagu** [comment, marked as answer]
While the behavior you've shown is definitely a bug, I think what you want is `SerializeAsAny`, not `InstanceOf`:
```python
from pydantic import BaseModel, SerializeAsAny

class UserEventModel(BaseModel):
    pass

class TestEventModel(UserEventModel):
    test: str


class WebhookPayload(BaseModel):
    payload: SerializeAsAny[UserEventModel]


print(WebhookPayload(payload=TestEventModel(test='bla')).model_dump())
#> {'payload': {'test': 'bla'}}
print(WebhookPayload(payload=TestEventModel(test='bla')).model_dump_json())
#> {"payload":{"test":"bla"}}
```
If for some reason you need `InstanceOf` (e.g., if `UserEventModel` wasn't actually a `BaseModel` subclass or whatever), you could also do `InstanceOf[SerializeAsAny[UserEventModel]]`. But I think for your use case here `SerializeAsAny` is the right thing.

I plan to fix the issue for `InstanceOf`, but the fix will likely result in both of them excluding the field in your example, rather than including it. (And you'd use `SerializeAsAny` to get it included.)

**Reply by dmontagu** [reply]
Opened a PR to fix the bug: https://github.com/pydantic/pydantic/pull/6829 — note this will cause the code above to behave differently, and the right solution will be to use `SerializeAsAny` to keep the extra field info in both calls.

**Reply by fproske** [reply]
Thanks, @dmontagu! `SerializeAsAny` was indeed what I was looking for and now I could also find the documentation for this. It works as intended and I'm happy I could help find a bug in the process. Thanks for the amazing work you guys do and best regards!
